[LiveComponent] add LiveProp name to modifier function

jannes-io opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Docs?         | yes
| Issues        | Fix #2650
| License       | MIT

Implements proposal 2 from the examples in the RFC.

Personally I would just pass the entire liveprop metadata object to this function instead of only the name, but it's marked internal so I'm not sure if we should be exposing it to users like that.

Commit:
[Notify] Added body, icon, tag ,renotify to controller.js for mercure-notifier

Co-authored-by: Simon André <smn.andre@gmail.com>

Fan2Shrek committed:
[Turbo] Add support for authentication to the EventSource via `turbo_stream_listen`

seb-jean committed:
Add Twig Extensions for page refreshes

Bartheyrman22 committed:
[Autocomplete] Escape `querySelector` dynamic selector with `CSS.escape()`

stehled committed:
[LiveComponent] `ComponentWithFormTrait` now correctly checks for a `TranslatableInterface` placeholder for `<select>` elements, close #2621

Co-authored-by: Simon André <smn.andre@gmail.com>

Halleck45 committed:
[Toolkit] Adds engine for parsing, downloading and manage components

Halleck45 committed:
[Toolkit] Adds first components

Halleck45 committed:
[Toolkit] Updated documentation in order to preview components

jannes-io committed:
[LiveComponent] add LiveProp name to modifier function

norkunas committed:
[LiveComponent] Add uid support for hydration

Kocal committed:
[Map] Make renderer tests way easier to maintain, use snapshots

Enz000 committed:
[Icons] Add xmlns attribute to svg icons

Kocal committed:
[Toolkit] Fine-tune the documentation

Kocal committed:
[Toolkit] Remove "Grid" component

See https://github.com/symfony/ux/pull/2539#discussion_r1993149852

Kocal committed:
[Toolkit] Open dependencies constraints

Kocal committed:
[Toolkit] Fix UXToolkitBundle name, refactor DependencyInjection/* in UXToolkitBundle,

dbu committed:
improve setup instructions

i missed that i need the http-client installed, the bundle silently reduces functionality without it. make it more clear what needs to be installed

smnandre commented on src/LiveComponent/doc/index.rst:
```diff
@@ -2617,7 +2617,9 @@ definition::
 Then the ``query`` value will appear in the URL like ``https://my.domain/search?q=my+query+string``.
 
-If you need to change the parameter name on a specific page, you can leverage the :ref:`modifier <modifier>` option::
+If you need to change the parameter name on a specific page, you can leverage the :ref:`modifier <modifier>` option
```
A couple of remarks here (unrelated to wether we will move forward with this PR or not :) )

First of all: thank you for thinking immediately "documentation". That's a very nice habit to have. 

Now for the changes here: I think we should not _change_ the existing documentation, but instead we may _add_ a dedicated paragraph _after_ this part. 

The current version does it jobs, covers the feature and how most developers use it, while keeping things small / easy to understand. 

Moreover, using a prefix instead of an alias here would offer less possibilities than before, so we should probably not describe this solution in first examples, but only when a prefix _could_ be justified (many identical components in the same page ? still not sure to be honest)

So here I think a dedicated paragraph after this one, for more advanced use case, could be more appropriate.

smnandre commented on src/LiveComponent/src/Metadata/LivePropMetadata.php:
```diff
@@ -135,7 +135,7 @@ public function withModifier(object $component): self
         }
 
-        $modifiedLiveProp = $component->{$modifier}($this->liveProp);
+        $modifiedLiveProp = $component->{$modifier}($this->liveProp, $this->getName());
```
For a change like this, a test would be needed. Especially for things related to query, URL, etc.

(The fact tests do pass after your changes does not proves it does work (nor that our current suite of test is good enough, to be fair))

jannes-io replied:
Added to the existing test, I don't see the need to make a separate test for just this param unless there are some edge cases that aren't entirely obvious to me at this time.

smnandre replied:
Let's test a page with two instances of same component, different prefix used in the modifier ... and check that the correct one as the good value and not the other :)

Kocal committed:
[Toolkit] Big refactoring about the registry & kits system, kits architecture, and value-objects

- The registry system and JSON serialization were fully reworked, no more useless compilation to JSON
- Rename "default" theme to "shadcn"
- The "theme" term is now known as "kit"
useless JSON serialization
- Abuse ValueObjects usage for File, Example, Dependency, ...
 - Add binary for vendors

Kocal committed:
[Toolkit] Rework the documentation

zak39 committed:
chore: Adjust changelog for features moving to 2.24 release

The following features were moved in the changelog as they are not present in the 2.23 release, but will be included in the 2.24 release

-  `id` to `Marker`, `Polygon` and `Polyline` constructors
-  `Map::removeMarker(string|Marker $markerOrId)`
-  `Map::removePolygon(string|Polygon $polygonOrId)`
-  `Map::removePolyline(string|Polyline $polylineOrId)`

weaverryan commented:
I personally don't have a big issue with removing the internal. Very thoughtful to realize that issue :).

Alternatively, if there is something you need on the metadata, we could add it to LiveProp

smnandre commented:
I don't see a problem to pass the second argument  👍 
(we should precise in docs this information may not be the original name of the prop in the class)

Passing the metadata is not something I'd love to do, without an absolute need to do so.
Anything we expose cannot then be changed afterwards.. And it did bring us in situations were we're a bit "stuck" to move forward / improve features.

Kocal committed:
Update versions to 2.24.0

kbond committed:
remove banner

kbond committed:
[Icons] improve DX when `symfony/http-client` is not installed

- throw exception if http-client is not available
- always enable iconify, iconify commands and on-demand registry
- display "potential" missing icons to `ux:icons:warm-cache`

Kocal committed:
[CI] Fix version bump for npm packages

jannes-io commented:
> I don't see a problem to pass the second argument 👍 (we should precise in docs this information may not be the original name of the prop in the class)

Okidoki, I can do the requested changes to the docs and add some tests to finish off this PR. Could you give me a case where the property name could be different from the name in the metadata? Or where I could find code that could create this side-effect so I can try to reproduce? Just want to understand what the implications might be there.

> Passing the metadata is not something I'd love to do, without an absolute need to do so. Anything we expose cannot then be changed afterwards.. And it did bring us in situations were we're a bit "stuck" to move forward / improve features.

Agreed 😄

jannes-io commented on src/LiveComponent/doc/index.rst:
```diff
@@ -2651,6 +2651,47 @@ This way you can also use the component multiple times in the same page and avoi
     <twig:SearchModule alias="q1" />
     <twig:SearchModule alias="q2" />
 
+.. versionadded:: 2.25
+
+   The property name is passed into the modifier function since LiveComponents 2.25.
```
This'd have to be updated to the actual release version, I just took the next one.

Kocal committed:
[CI] Adjust tags to "v2.*.*" for NPM auto-releases

smnandre committed:
[CI] Use git diff instead of git status

The check failed here because file changed .. but with no content change.

Kocal committed:
[Toolkit] Rework all components for Shadcn kit, add more, remove some, add docs too

Kocal committed:
Add .editorconfig, enforce LF line-ending style

seb-jean committed:
[Turbo] Minor documentation syntax fixes

Fan2Shrek committed:
Pass topics to mercure if not defined

Kocal committed:
[Map] Downgrade PHP requirement from 8.3 to 8.1

Kocal committed:
[Toolkit] Remove tales-from-a-dev/twig-tailwind-extra dependency, to allow PHP 8.1 minimal support

Kocal committed:
[Toolkit] Refactor `ToolkitKit` enum to `ToolkitKitId`, leverage description/uxIcon/installation steps in `Kit` VO

Kocal committed:
[Docs] Add step to disable `package.json` synchronization while upgrading PHP package

Kocal committed:
[Toolkit] Add StimulusController support, extract logic to KitSynchronizer, rewrite installation system, create dedicated Installer/Pool system, make classes not readonly (PHP 8.1 comptability)

Kocal committed:
[Toolkit] Update version handling in dependencies to use string format, improve command output messages, improve InstallComponentCommand (interactive ask), remove LintKitCommand

Kocal committed:
[Toolkit] Do not use FileSystem::readFile(), as it has been implemented in Symfony 7.1 and we want to support at least Symfony 6.4

Kocal committed:
[Toolkit] Improve InstallComponentCommand by asking/guessing which Kit to use, remove ux_toolkit.kit parameter, remove DependencyInjection configuration

jannes-io commented:
@smnandre , @weaverryan Is there anything I can do to move this along?

Kocal committed:
[StimulusBundle] Make the JS package private

smnandre commented on src/LiveComponent/doc/index.rst:
```diff
@@ -2651,6 +2651,47 @@ This way you can also use the component multiple times in the same page and avoi
+The ``modifier`` function can also take the name of the property as a secondary parameter.
+It can be used to perform more generic operations inside of the modifier that can be re-used for multiple props::
+
+    abstract class AbstractSearchModule
+    {
+        #[LiveProp(writable: true, url: true, modifier: 'modifyQueryProp')]
+        public string $query = '';
+
+        protected string $urlPrefix = '';
+
+        public function modifyQueryProp(LiveProp $liveProp, string $propName): LiveProp
+        {
+            if ($this->urlPrefix) {
+                return $liveProp->withUrl(new UrlMapping(as: $this->urlPrefix.'-'.$propName));
+            }
+            return $liveProp;
+        }
+    }
+
+    #[AsLiveComponent]
+    class ImportantSearchModule extends AbstractSearchModule
+    {
+    }
+
+    #[AsLiveComponent]
+    class SecondarySearchModule extends AbstractSearchModule
+    {
+        protected string $urlPrefix = 'secondary';
+    }
```
I'm wondering.. could we find use an example more focused on the feature than the PHP code structure abstract/extends here ?

I was thinking something like "imagine you have a CMS theme editor with a split view allowing you to adjust your design both in light and dark mode. ... 

You could use a ColorTheme LiveComponent for both, but then [describe problem]

```
<twig:ColorSettings prefix="dark" color="..." />

<twig:ColorSettings prefix="light" color="..." />

```

Thanks to the second parameter of YYY, you can dynamically prefix [...].
 
 ```        
     #[LiveProp]
      public string $theme;
 
      #[LiveProp(url: true, modifier: 'prefixParam')]
      public string $color;

        public function prefixParam($liveProp, string $propName): LiveProp
        {
                return $liveProp->withUrl(
                   // Light Theme will use ?light-color
                    new UrlMapping(as: $this->prefix.'-'.$propName)
               );
        }
```

Now your query parameters will be: XXX ZZZ 

It even works with the as: property ..  

```php

     // query param will be ZZZ

      #[LiveProp(url: true, as: 'bg' modifier: 'prefixParam')]
      public string $background;      
```




wdyt ?

Kocal committed:
[Site] Rename CodePreview_Tabs to Toolkit_Tabs, extract Toolkit's documentation tabs rendering into ToolkitService

Kocal committed:
[Site] Add manual installation steps for Toolkit kits's components

Kocal committed:
[Toolkit] Remove redundant code in Shadcn kit components

smnandre committed:
[TwigComponent] Fix `loadTemplate` deprecation for Twig >= 3.21

Fix #2710

cf https://github.com/twigphp/Twig/pull/4583

smnandre committed:
[Toolkit] Fix deprecation Kernel > 7.3

Kocal committed:
[StimulusBundle] Make the JS package private (for real)

Kocal committed:
[Site] Fix Terminal component styles, ensure it does not grow, and display scrollbars

Kocal committed:
[Site] Refactor and use the CodeBlockRenderer::highlightCode() method to render the highlighted code inside a Terminal component

smnandre committed:
[ToolKit] Add missing Interface suffixes

Fix #2539

Kocal committed:
[Toolkit] Introduce KitContextRunner, to run code in the context of a given Kit

Kocal committed:
[Toolkit] Add functional tests to render all Kit components usage codes (from their documentation), with a snapshot system

Kocal committed:
[Toolkit] Remove documentation about non-existant component AlertDialog (will be re-added later 😇)

Kocal committed:
[Toolkit] Pin symfony/phpunit-bridge to ^7.2 and phpunit to ^9.6.22

Kocal committed:
[Toolkit] Remove checks layers for tales-from-a-dev/twig-tailwind-extra on PHP < 8.2, since it now supports PHP 8.1

Kocal committed:
[Toolkit] Improve descriptions of Shadcn components

Kocal committed:
[Toolkit] Minor adjustements on Kit creation command (reword questions, fix doc format)

Kocal committed:
[Toolkit] Remove Kit "authors"

nikophil committed:
fix: allow LiveComponentHydrator::hydrateValue() to hydrate null values

Kocal committed:
[Toolkit] Update snapshots & fix tests, following #2728

smnandre commented:
LGTM, and we will improve the docs afterwards

(changelog must be updated, though)

jannes-io commented:
> (changelog must be updated, though)

Done.

As for the other remarks, apologies for not getting back to you on that.

1. For the test, I didn't find an existing test where actual LiveComponents are rendered on a page. I'm afraid I'm a little out of my depth to be the first to produce such a test. If there is such a test that can be easily copied and modified to test the new parameter, and I just missed it, I can take a look.
2. For the docs, it feels to me like the feature is adequately technically documented, we don't look at docs to read a story, we're there for information on how to use features so we can use them in our own stories. If someone wants to rewrite it with a better example, that's fair 😆

Thanks for your assistance.

jannes-io commented:
What in tarnation happened here... Fabbot told me not to merge, so I rebased, and now all this crap is in here..

CBA to fiddle with git, remade it here: #2744
